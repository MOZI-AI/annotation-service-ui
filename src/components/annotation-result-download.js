import React from "react";
import logo from "../assets/mozi_globe.png";
import { Grid, Button } from "@material-ui/core";
import { ArrowDownward } from "@material-ui/icons";

export class AnnotationResultDownload extends React.Component {
  render() {
    return (
      <React.Fragment>
        <Grid containe style={{ paddingTop: "30px" }}>
          <Grid item style={{ textAlign: "center" }}>
            <img alt="MOZI globe logo" src={logo} style={{ width: "100px" }} />
            <h2>Annotaion result ready!</h2>
            <p>
              The result was too large for visualization. Please click the
              button below to download the annotation result file.{" "}
            </p>
            <Button
              variant="contained"
              onClick={e => this.props.downloadSchemeFile()}
              color="primary"
            >
              <ArrowDownward />
              Download annotation results
            </Button>
          </Grid>
        </Grid>
      </React.Fragment>
    );
  }
}
