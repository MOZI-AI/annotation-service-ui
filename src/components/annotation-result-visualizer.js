import React from "react";
import {
  Alert,
  Collapse,
  Checkbox,
  Row,
  Col,
  Button,
  Icon,
  Tooltip
} from "antd";
import * as cytoscape from "cytoscape";
import { CYTOSCAPE_COLA_CONFIG, CYTOSCAPE_STYLE } from "../visualizer.config";
import * as cola from "cytoscape-cola";

const AnnotationColors = [
  "#D8E0F1",
  "#EBE1EE",
  "#F0EDD9",
  "#D9F7D9",
  "#EEE",
  "#F1D8D8"
];

export class AnnotationResultVisualizer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      selectedNode: null,
      history: []
    };

    this.cy_wrapper = React.createRef();
    cytoscape.use(cola);
  }

  changeLayout() {
    this.layout = !this.props.minimalMode
      ? this.cy.layout(CYTOSCAPE_COLA_CONFIG)
      : this.cy.layout({ name: "cose" });
    this.layout.run();
  }

  takeScreenshot() {
    const image = this.cy.jpg();
    const link = document.createElement("a");
    link.setAttribute("href", image);
    link.setAttribute("download", "mozi-graph.jpg");
    link.click();
  }

  componentDidMount() {
    this.cy = cytoscape({
      container: this.cy_wrapper.current,
      pixelRatio: 1,
      hideEdgesOnViewport: true,
      textureOnViewport: true
    });
    this.cy.startBatch();
    this.cy.add(this.props.graph.nodes.filter(e => e.data.group === "Gene"));
    this.cy.endBatch();
    this.layout = !this.props.minimalMode
      ? this.cy.layout(CYTOSCAPE_COLA_CONFIG)
      : this.cy.layout({ name: "grid" });
    this.cy.style(
      !this.props.minimalMode
        ? CYTOSCAPE_STYLE.concat(this.assignColorToAnnotations())
        : CYTOSCAPE_STYLE
    );
    this.registerEventListeners();
    this.layout.run();
  }

  registerEventListeners() {
    this.cy.nodes().on(
      "select",
      function(event) {
        this.focusOnNode(event.target.data().id);
        this.setState({ selectedNode: event.target.data() });
      }.bind(this)
    );
    this.cy.nodes().on(
      "unselect",
      function(event) {
        this.removeFocus();
        this.setState({ selectedNode: null });
      }.bind(this)
    );
  }

  removeFocus() {
    this.cy.json(this.graphStateBeforeFocus);
    this.cy.batch(() => {
      this.cy.nodes().style({ opacity: 1 });
      this.cy.edges().style({ opacity: 1 });
    });
  }

  assignColorToAnnotations() {
    return this.props.annotations.reduce((acc, ann, i, arr) => {
      acc.push({
        selector: 'edge[group="' + ann + '"]',
        style: { "line-color": AnnotationColors[i] }
      });
      return acc;
    }, []);
  }

  focusOnNode(id) {
    this.graphStateBeforeFocus = this.cy.json();
    const hood = this.cy.getElementById(id).closedNeighborhood();

    this.cy.batch(() => {
      this.cy.delay(100, () => {
        hood.layout({ name: "concentric", fit: true }).run();
      });
      this.cy
        .nodes()
        .difference(hood)
        .style({ opacity: 0.1 });
      this.cy
        .edges()
        .difference(hood)
        .style({ opacity: 0.1 });
    });
  }

  toggleAnnotationVisibility(annotation, show) {
    show
      ? this.cy.batch(() => {
          this.cy.add(
            this.props.graph.nodes.filter(e => e.data.group === annotation)
          );
          this.cy.add(
            this.props.graph.edges.filter(e => e.data.group === annotation)
          );
        })
      : this.cy.remove(`[group='${annotation}']`);
    this.changeLayout();
    this.registerEventListeners();
  }

  render() {
    return (
      <div
        style={{
          minHeight: "90vh"
        }}
      >
        <Row>
          <Col span={1}>
            <Button.Group
              style={{
                position: "absolute",
                display: "flex",
                flexDirection: "column"
              }}
              size="large"
            >
              <Tooltip placement="right" title="Change layout">
                <Button
                  onClick={e => this.changeLayout()}
                  style={{ border: "none" }}
                >
                  <Icon type="swap" />
                </Button>
              </Tooltip>
              <Tooltip placement="right" title="Save screenshot">
                <Button
                  onClick={e => this.takeScreenshot()}
                  style={{ border: "none" }}
                >
                  <Icon type="camera" />
                </Button>
              </Tooltip>

              {this.state.history.length > 0 && (
                <Tooltip placement="right" title="Undo">
                  <Button
                    style={{ border: "none", borderRadius: "0" }}
                    onClick={e => this.undo()}
                  >
                    <Icon type="undo" />
                  </Button>
                </Tooltip>
              )}

              <Tooltip placement="right" title="Download scheme file">
                <Button
                  onClick={e => this.props.downloadFile()}
                  ghost
                  type="primary"
                  style={{ border: "none", borderRadius: "0" }}
                >
                  <Icon type="download" />
                </Button>
              </Tooltip>
            </Button.Group>
          </Col>
          <Col span={19}>
            <div>
              <div
                style={{
                  minHeight: "90vh"
                }}
                ref={this.cy_wrapper}
              />
            </div>
          </Col>
          <Col span={4}>
            <Collapse
              accordion
              bordered={false}
              defaultActiveKey="2"
              style={{
                background: "none",
                width: "250px"
              }}
            >
              <Collapse.Panel header="Annotations" key="2">
                {this.props.annotations.map((a, i) => (
                  <React.Fragment key={a}>
                    <Checkbox
                      style={
                        this.props.minimalMode
                          ? {}
                          : { backgroundColor: AnnotationColors[i] }
                      }
                      defaultChecked={false}
                      onChange={e =>
                        this.toggleAnnotationVisibility(a, e.target.checked)
                      }
                    >
                      {a}
                    </Checkbox>
                    <br />
                  </React.Fragment>
                ))}
              </Collapse.Panel>
            </Collapse>
          </Col>
        </Row>
        {this.state.selectedNode && (
          <Alert
            style={{
              position: "absolute",
              bottom: "15px",
              left: "15px",
              width: "350px",
              backgroundColor: "#c9e1f9",
              border: "solid 1px #87BEF5"
            }}
            message={`${this.state.selectedNode.name} ( ${
              this.state.selectedNode.id
            } )`}
            description={this.state.selectedNode.definition}
            closable
            onClose={() => {}}
          />
        )}
      </div>
    );
  }
}
