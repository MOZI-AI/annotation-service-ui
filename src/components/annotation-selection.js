import React from "react";

import { Checkbox, FormControlLabel } from "@material-ui/core";

const filterFormStyle = {
  paddingLeft: "25px",
  marginLeft: "5px",
  borderLeft: "dashed 1px #ddd"
};

export class AnnotationSelection extends React.Component {
  isAnnotationSelected = annotation =>
    this.props.selectedAnnotations.some(sa => sa.name === annotation);

  render() {
    return (
      <React.Fragment>
        <h4>Annotations: </h4>

        {this.props.availableAnnotations.map(a => (
          <React.Fragment key={a.key}>
            <FormControlLabel
              value="0"
              control={
                <Checkbox
                  name={a.key}
                  onChange={e =>
                    this.props.handleAnnotationsChanged(
                      e.target.checked,
                      e.target.name
                    )
                  }
                />
              }
              label={a.name}
            />
            <br />
            {this.isAnnotationSelected(a.key) && (
              <div style={filterFormStyle} layout="inline">
                {a.fitlerForm(a.defaults, filter =>
                  this.props.handleFilterChanged(a.key, filter)
                )}
              </div>
            )}
          </React.Fragment>
        ))}
      </React.Fragment>
    );
  }
}
