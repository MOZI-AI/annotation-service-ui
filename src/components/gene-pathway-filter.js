import React from "react";
import { Checkbox, Switch, Form, Row, Col } from "antd";
import FormItem from "antd/lib/form/FormItem";

const options = [
  { label: "SMPDB", value: "smpdb" },
  { label: "Reactome", value: "reactome" }
];

export class GenePathwayFilter extends React.Component {
  render() {
    return (
      <React.Fragment>
        <Form layout="inline">
          <FormItem label="Pathway">
            <Checkbox.Group
              options={options}
              defaultValue={this.props.defaults.namespace}
              onChange={e => this.props.handleFilterChanged({ namespace: e })}
            />
          </FormItem>

          <Row gutter={16}>
            <Col md={12} lg={8} xl={6}>
              <FormItem>
                <Switch
                  onChange={e => {
                    this.props.handleFilterChanged({
                      include_small_molecule: e
                    });
                  }}
                />{" "}
                Small molecules
              </FormItem>
            </Col>
            <Col md={12} lg={8} xl={6}>
              <FormItem>
                <Switch
                  onChange={e => {
                    this.props.handleFilterChanged({ include_prot: e });
                  }}
                />{" "}
                Proteins
              </FormItem>
            </Col>
          </Row>
        </Form>
      </React.Fragment>
    );
  }
}
