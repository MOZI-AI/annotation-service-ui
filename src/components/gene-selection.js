import React from "react";
import { Form, Input, Upload, Icon, Tag, Alert, Radio, Button } from "antd";

export const InputMethods = {
  DIRECT_INPUT: 1,
  FILE_UPLOAD: 2
};

export class GeneSelection extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      inputMethod: InputMethods.DIRECT_INPUT,
      currentGeneName: ""
    };
    this.uploaderAttributes = {
      name: "geneList",
      multiple: false,
      beforeUpload: geneList => {
        this.props.onGeneListUploaded(geneList);
        return false;
      }
    };
    this.handleGeneAdded = this.handleGeneAdded.bind(this);
  }

  handleGeneAdded() {
    if (!this.props.form.getFieldError("gene") && this.state.currentGeneName) {
      this.props.onGeneAdded(this.state.currentGeneName);
      this.props.form.setFieldsValue({ gene: "" });
      this.setState({ currentGeneName: "" });
    }
  }

  render() {
    const { getFieldDecorator } = this.props.form;

    return (
      <React.Fragment>
        <Form>
          <Form.Item>
            <Radio.Group
              id="inputMethod"
              onChange={e => this.setState({ inputMethod: e.target.value })}
              value={this.state.inputMethod}
            >
              <Radio value={InputMethods.DIRECT_INPUT}>Input directly</Radio>
              <Radio value={InputMethods.FILE_UPLOAD}>Import from file</Radio>
            </Radio.Group>
          </Form.Item>
        </Form>

        {this.state.inputMethod === InputMethods.DIRECT_INPUT && (
          <Form id="directInputForm">
            <Form.Item label="Gene name" style={{ marginBottom: "0" }}>
              {getFieldDecorator("gene", {
                rules: [
                  { required: true, message: "Please input gene name!" },
                  {
                    validator: (rule, value, callback) => {
                      if (this.props.genes.indexOf(value) !== -1) {
                        callback("This gene name is already added.");
                      }
                      callback();
                    }
                  }
                ]
              })(
                <Input
                  suffix={<Icon type="enter" />}
                  placeholder="Input gene name and hit enter"
                  onChange={e =>
                    this.setState({ currentGeneName: e.target.value.trim() })
                  }
                  onPressEnter={() => {
                    this.handleGeneAdded();
                  }}
                />
              )}
            </Form.Item>
          </Form>
        )}

        {this.state.inputMethod === InputMethods.FILE_UPLOAD && (
          <React.Fragment>
            <Upload.Dragger {...this.uploaderAttributes} fileList={[]}>
              <p className="ant-upload-drag-icon">
                <Icon type="import" />
              </p>
              <p className="ant-upload-text">
                Click here or drag file containing list of genes.
              </p>
              <p className="ant-upload-hint">
                Please make sure the file is a plain text file containing a gene
                name per line.
              </p>
            </Upload.Dragger>

            {this.props.geneList && this.props.genes.length ? (
              <Alert
                className="fileDetails"
                style={{
                  borderRadius: 0,
                  borderTop: "none",
                  position: "relative",
                  bottom: "2px"
                }}
                message={
                  this.props.geneList.name +
                  " ( " +
                  this.props.geneList.size / 1000000 +
                  " MB )"
                }
                type="info"
              />
            ) : null}
          </React.Fragment>
        )}

        <div style={{ marginTop: "10px" }}>
          <h4 style={{ color: "#82909d", marginBottom: "5px" }}>
            {" "}
            Selected genes
            {this.props.genes.length ? (
              <Button
                onClick={e => this.props.onAllGenesRemoved()}
                type="danger"
                ghost
                style={{ border: "none" }}
              >
                {" "}
                Remove all
              </Button>
            ) : null}
          </h4>
          {!this.props.genes.length ? (
            <Alert
              id="noGenesAlert"
              type="warning"
              description="The genes you input appear here. You may input gene names directly or import them from file."
            />
          ) : null}

          {this.props.genes.length
            ? this.props.genes.map(gene => (
                <Tag
                  style={{ marginBottom: "5px" }}
                  onClose={() => this.props.onGeneRemoved(gene)}
                  key={gene}
                  closable
                  color="blue"
                >
                  {" "}
                  {gene}{" "}
                </Tag>
              ))
            : null}
        </div>
      </React.Fragment>
    );
  }
}

export const GeneSelectionForm = Form.create()(GeneSelection);
