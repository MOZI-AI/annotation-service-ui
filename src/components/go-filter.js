import React from "react";
import { Checkbox, Switch, InputNumber, Form, Row, Col } from "antd";
import FormItem from "antd/lib/form/FormItem";

const namespaces = [
  { label: "Biological process", value: "biological_process" },
  { label: "Cellular component", value: "cellular_component" },
  { label: "Molecular function", value: "molecular_function" }
];

const formItemStyle = {
  marginBottom: 5
};

export class GOFilter extends React.Component {
  render() {
    return (
      <React.Fragment>
        <Form layout="inline">
          <FormItem label="Namespace">
            <Checkbox.Group
              onChange={namespaces => {
                this.props.handleFilterChanged({ namespace: namespaces });
              }}
              options={namespaces}
              defaultValue={this.props.defaults.namespace}
            />
          </FormItem>

          <Row gutter={16}>
            <Col md={24} lg={12} xl={8}>
              <FormItem label="Number of parent terms" style={formItemStyle}>
                <InputNumber
                  min={0}
                  defaultValue={this.props.defaults.parents}
                  onChange={n => this.props.handleFilterChanged({ parents: n })}
                />
              </FormItem>
            </Col>
          </Row>
          <Row gutter={16}>
            <Col md={12} lg={8} xl={6}>
              <FormItem>
                <Switch
                  defaultChecked={this.props.defaults.get_entrez_id}
                  onChange={e => {
                    this.props.handleFilterChanged({ get_entrez_id: e });
                  }}
                />{" "}
                Include Entrez ID
              </FormItem>
            </Col>
          </Row>
        </Form>
      </React.Fragment>
    );
  }
}
