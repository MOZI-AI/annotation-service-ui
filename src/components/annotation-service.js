import React from "react";
import { SERVER_ADDRESS, showNotification } from "../utils";
import { Annotate } from "../proto/annotation_pb_service";
import {
  AnnotationRequest,
  Annotation,
  Gene,
  Filter
} from "../proto/annotation_pb";
import { grpc } from "grpc-web-client";
import { GeneSelectionForm } from "./gene-selection";
import { AnnotationSelection } from "./annotation-selection";
import { Button, Grid } from "@material-ui/core";
import { Check, CheckCircle } from "@material-ui/icons";

export class AnnotationService extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      genes: [],
      geneList: null,
      selectedAnnotations: [],
      annotationResult: null,
      busy: false,
      notification: null
    };
    // bind functions
    this.handleGeneAdded = this.handleGeneAdded.bind(this);
    this.handleGeneRemoved = this.handleGeneRemoved.bind(this);
    this.handleGeneListUploaded = this.handleGeneListUploaded.bind(this);
    this.handleAllGenesRemoved = this.handleAllGenesRemoved.bind(this);
    this.handleAnnotationsChanged = this.handleAnnotationsChanged.bind(this);
    this.handleFilterChanged = this.handleFilterChanged.bind(this);
    this.downloadSchemeFile = this.downloadSchemeFile.bind(this);
  }

  handleGeneAdded(input) {
    this.setState(state => {
      let genes = state.genes.slice(0);
      genes = [
        ...genes,
        ...input
          .trim()
          .toUpperCase()
          .split(" ")
      ].filter((g, i, arr) => g && arr.indexOf(g) === i);
      return { genes: genes };
    });
  }

  handleGeneRemoved(gene) {
    this.setState(state => {
      let genes = state.genes.slice(0);
      return { genes: genes.filter(g => g !== gene) };
    });
  }

  handleGeneListUploaded(geneList) {
    const fileReader = new FileReader();
    fileReader.readAsText(geneList);
    fileReader.onload = () => {
      const re = /^[a-z0-9\s]+$/i;
      if (!re.test(fileReader.result)) {
        this.setState({
          notification: {
            message: "The selected file contains invalid characters.",
            busy: false
          }
        });
        return;
      }
      const geneArray = fileReader.result.split("\n");
      const uniqueGeneArray = geneArray.reduce((accumulator, value) => {
        if (value && accumulator.indexOf(value) === -1) accumulator.push(value);
        return accumulator;
      }, []);
      this.setState({ geneList: geneList, genes: uniqueGeneArray });
    };
  }

  handleAllGenesRemoved() {
    this.setState({ genes: [], geneList: null });
  }

  handleAnnotationsChanged(isSelected, annotation) {
    this.setState(state => {
      let selectedAnnotations = state.selectedAnnotations.slice();
      isSelected
        ? selectedAnnotations.push({
            name: annotation,
            filter: this.props.availableAnnotations.find(
              a => a.key === annotation
            ).defaults
          })
        : (selectedAnnotations = selectedAnnotations.filter(
            a => a.name !== annotation
          ));

      return { selectedAnnotations: selectedAnnotations };
    });
  }

  handleFilterChanged(annotation, filter) {
    this.setState(state => {
      const selectedAnnotations = state.selectedAnnotations.map(sa => {
        if (sa.name === annotation) {
          sa.filter = Object.assign({}, sa.filter, filter);
        }
        return sa;
      });
      return { selectedAnnotations: selectedAnnotations };
    });
  }

  isFormValid() {
    let valid = true;
    valid = valid && this.state.selectedAnnotations.length;
    valid = valid && this.state.genes.length;
    // If Gene GO annotation is selected, namespace must be defined
    const GO = this.state.selectedAnnotations.find(
      a => a.name === "gene_go_annotation"
    );
    if (GO) valid = valid && GO.filter.namespace.length;
    // If Gene Pathway annotation is selected, namespace must be defined
    const Pathway = this.state.selectedAnnotations.find(
      a => a.name === "gene_pathway_annotation"
    );
    if (Pathway) {
      valid =
        valid &&
        (Pathway.filter.namespace.includes("smpdb") ||
          Pathway.filter.namespace.includes("reactome"));
    }
    return valid;
  }

  downloadSchemeFile() {
    const json = `data:application/txt, ${encodeURIComponent(
      this.state.annotationResult.schemeFile
    )}`;
    const link = document.createElement("a");
    link.setAttribute("href", json);
    link.setAttribute("download", "annotations-scheme.scm");
    link.click();
  }

  capitalizeFirstLetter(string) {
    return string.charAt(0).toUpperCase() + string.slice(1);
  }

  handleSubmit() {
    const annotationResult = new AnnotationRequest();
    annotationResult.setGenesList(
      this.state.genes.map(g => {
        const gene = new Gene();
        gene.setGenename(g);
        return gene;
      })
    );
    annotationResult.setAnnotationsList(
      this.state.selectedAnnotations.map(sa => {
        const annotation = new Annotation();
        annotation.setFunctionname(sa.name);
        annotation.setFiltersList(
          sa.filter
            ? Object.keys(sa.filter).map(k => {
                const filter = new Filter();
                filter.setFilter(k);
                filter.setValue(
                  Array.isArray(sa.filter[k])
                    ? sa.filter[k]
                        .reduce((acc, value) => {
                          return acc + " " + value;
                        }, "")
                        .trim()
                    : this.capitalizeFirstLetter(sa.filter[k].toString())
                );
                return filter;
              })
            : []
        );
        return annotation;
      })
    );

    this.setState({
      notification: { message: "Fetching annotation results ...", busy: true }
    });

    grpc.unary(Annotate.Annotate, {
      request: annotationResult,
      host: SERVER_ADDRESS,
      onEnd: ({ status, statusMessage, message }) => {
        if (status === grpc.Code.OK) {
          this.setState(state => ({
            busy: false,
            annotationResult: message.array[0],
            notification: null
          }));
        } else {
          if (statusMessage.includes("Gene Doesn't exist")) {
            const invalidGenes = statusMessage
              .split("`")[1]
              .split(",")
              .map(g => g.trim())
              .filter(g => g);
            this.setState(state => ({
              busy: false,
              genes: state.genes.filter(g => !invalidGenes.includes(g)),
              notification: { message: statusMessage, busy: false }
            }));
          } else {
            this.setState(state => ({
              busy: false,
              notification: { message: statusMessage, busy: false }
            }));
          }
        }
      }
    });
  }

  render() {
    return (
      <React.Fragment>
        {!this.state.annotationResult && (
          <div>
            {this.state.notification &&
              showNotification(this.state.notification, () => {
                this.setState({ notification: null });
              })}
            <GeneSelectionForm
              genes={this.state.genes}
              geneList={this.state.geneList}
              onGeneAdded={this.handleGeneAdded}
              onGeneRemoved={this.handleGeneRemoved}
              onGeneListUploaded={this.handleGeneListUploaded}
              onAllGenesRemoved={this.handleAllGenesRemoved}
            />
            <AnnotationSelection
              handleAnnotationsChanged={this.handleAnnotationsChanged}
              handleFilterChanged={this.handleFilterChanged}
              selectedAnnotations={this.state.selectedAnnotations}
              availableAnnotations={this.props.availableAnnotations}
            />
            <Grid container justify="flex-end">
              <Grid item>
                <Button
                  id="submitButton"
                  variant="contained"
                  onClick={() => this.handleSubmit()}
                  disabled={!this.isFormValid()}
                  color="primary"
                >
                  <Check />
                  Submit
                </Button>
              </Grid>
            </Grid>
          </div>
        )}
        {this.state.annotationResult ? (
          <Grid container justify="center">
            <Grid
              item
              style={{
                textAlign: "center",
                display: "flex",
                flexDirection: "column",
                alignItems: "center",
                justifyContent: "center",
                height: "60vh"
              }}
            >
              <CheckCircle style={{ fontSize: "72px", color: "#54C21F" }} />
              <h1 style={{ margin: 5 }}>Your request is being processed</h1>
              <h3 style={{ marginTop: 0, color: "#555" }}>
                Follow the link below to view results.
              </h3>
              <a
                rel="noopener noreferrer"
                target="_blank"
                href={this.state.annotationResult}
                style={{ fontSize: 18 }}
              >
                {this.state.annotationResult}
              </a>
            </Grid>
          </Grid>
        ) : null}
      </React.Fragment>
    );
  }
}
