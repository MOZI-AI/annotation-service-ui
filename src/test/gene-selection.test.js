import React from "react";
import { mount, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { GeneSelectionForm } from "../components/gene-selection";
import renderer from "react-test-renderer";

configure({ adapter: new Adapter() });

describe("<GeneSelectionForm />", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<GeneSelectionForm genes={[]} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("shows direct input form initially", () => {
    const wrapper = mount(<GeneSelectionForm genes={[]} />);
    expect(wrapper.find("#directInputForm").exists()).toBeTruthy();
  });

  it("shows an alert message for empty gene list", () => {
    const wrapper = mount(<GeneSelectionForm genes={[]} />);
    expect(wrapper.find("Alert#noGenesAlert").exists()).toBeTruthy();
  });

  it("renders a list item for each inserted gene", () => {
    const wrapper = mount(<GeneSelectionForm genes={["G1", "G2", "G3"]} />);
    expect(wrapper.find("Tag").length).toEqual(3);
  });
});
