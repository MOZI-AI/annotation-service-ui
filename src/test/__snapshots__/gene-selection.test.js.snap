// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GeneSelectionForm /> renders correctly 1`] = `
Array [
  <div
    className="MuiFormGroup-root-1"
    defaultValue="1"
    id="inputMethod"
    role="radiogroup"
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
      }
    }
  >
    <label
      className="MuiFormControlLabel-root-3"
    >
      <span
        className="MuiButtonBase-root-24 MuiIconButton-root-18 MuiPrivateSwitchBase-root-14 MuiRadio-root-9 MuiRadio-colorSecondary-13 MuiPrivateSwitchBase-checked-15 MuiRadio-checked-10"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label-23"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-27"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zm0-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
            />
          </svg>
          <input
            checked={true}
            className="MuiPrivateSwitchBase-input-17"
            onChange={[Function]}
            type="radio"
            value="1"
          />
        </span>
        <span
          className="MuiTouchRipple-root-131"
        />
      </span>
      <span
        className="MuiTypography-root-36 MuiTypography-body1-45 MuiFormControlLabel-label-8"
      >
        Input directly
      </span>
    </label>
    <label
      className="MuiFormControlLabel-root-3"
    >
      <span
        className="MuiButtonBase-root-24 MuiIconButton-root-18 MuiPrivateSwitchBase-root-14 MuiRadio-root-9 MuiRadio-colorSecondary-13"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label-23"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-27"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
            />
          </svg>
          <input
            checked={false}
            className="MuiPrivateSwitchBase-input-17"
            onChange={[Function]}
            type="radio"
            value="2"
          />
        </span>
        <span
          className="MuiTouchRipple-root-131"
        />
      </span>
      <span
        className="MuiTypography-root-36 MuiTypography-body1-45 MuiFormControlLabel-label-8"
      >
        Import from file
      </span>
    </label>
  </div>,
  <div
    className="MuiFormControl-root-72 MuiFormControl-marginDense-74 MuiFormControl-fullWidth-75"
    onKeyDown={[Function]}
  >
    <label
      className="MuiFormLabel-root-87 MuiInputLabel-root-76 MuiInputLabel-formControl-81 MuiInputLabel-animated-84 MuiInputLabel-marginDense-82 MuiInputLabel-outlined-86"
      data-shrink={false}
      htmlFor="directInputForm"
    >
      Gene name
    </label>
    <div
      className="MuiInputBase-root-107 MuiOutlinedInput-root-94 MuiInputBase-fullWidth-116 MuiInputBase-formControl-108 MuiInputBase-marginDense-114 MuiInputBase-adornedEnd-112 MuiOutlinedInput-adornedEnd-98"
      onClick={[Function]}
    >
      <fieldset
        aria-hidden={true}
        className="MuiPrivateNotchedOutline-root-124 MuiOutlinedInput-notchedOutline-101"
        style={
          Object {
            "paddingLeft": 8,
          }
        }
      >
        <legend
          className="MuiPrivateNotchedOutline-legend-125"
          style={
            Object {
              "width": 0.01,
            }
          }
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "&#8203;",
              }
            }
          />
        </legend>
      </fieldset>
      <input
        aria-invalid={false}
        className="MuiInputBase-input-117 MuiOutlinedInput-input-102 MuiInputBase-inputMarginDense-118 MuiOutlinedInput-inputMarginDense-103 MuiInputBase-inputAdornedEnd-123 MuiOutlinedInput-inputAdornedEnd-106"
        disabled={false}
        id="directInputForm"
        name="gene"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Input gene name and hit enter"
        required={false}
        type="text"
        value=""
      />
      <div
        className="MuiInputAdornment-root-126 MuiInputAdornment-positionStart-128"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-27"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M19 7v4H5.83l3.58-3.59L8 6l-6 6 6 6 1.41-1.41L5.83 13H21V7z"
          />
        </svg>
      </div>
    </div>
  </div>,
  <div
    style={
      Object {
        "marginTop": "10px",
      }
    }
  >
    <h4
      style={
        Object {
          "color": "#000",
          "marginBottom": "5px",
        }
      }
    >
      Selected genes:
    </h4>
    <div
      id="noGenesAlert"
      style={
        Object {
          "backgroundColor": "#FFFBE5",
          "border": "solid 1px #ffe7bc",
          "borderRadius": "3px",
          "padding": "15px",
        }
      }
    >
      The genes you input appear here. You may input gene names directly or import them from file.
    </div>
  </div>,
]
`;
